crates/storage/src/pruning.rs

```diff
@@ -0,0 +1,86 @@
+use std::sync::Arc;
+
+use anyhow::Context;
+use pathfinder_common::{BlockHeader, BlockNumber};
+use tokio::sync::broadcast::Receiver;
+
+use crate::{BlockId, Storage, Transaction};
+
+/// Runs a task that prunes the blockchain history when the number of blocks
+/// exceeds a certain threshold.
+pub fn run_blockchain_pruning(
+    prune_storage: Storage,
+    block_header_rx: Receiver<Arc<BlockHeader>>,
+    num_blocks_kept: u64,
+) -> tokio::task::JoinHandle<anyhow::Result<()>> {
+    pub async fn blockchain_pruning(
+        prune_storage: Storage,
+        mut block_header_rx: Receiver<Arc<BlockHeader>>,
+        num_blocks_kept: u64,
+    ) -> anyhow::Result<()> {
+        let block_threshold = match num_blocks_kept {
+            0 => 1,
+            n => std::cmp::min(3 * n / 2, 1000),
+        };
+        tracing::info!(history_kept=%num_blocks_kept, "Blockchain pruning enabled");
+
+        let mut block_cnt = 0;
+
+        loop {
+            let header = block_header_rx.recv().await?;
+            block_cnt += 1;
+
+            if block_cnt >= block_threshold {
+                block_cnt = 0;
+                tokio::task::block_in_place(|| {
+                    prune_storage
+                        .connection()
+                        .context("Creating database connection for blockchain pruning")?
+                        .transaction_with_behavior(
+                            // Avoid contention with the transaction that writes the block header
+                            // data.
+                            crate::TransactionBehavior::Immediate,
+                        )
+                        .context("Creating database transaction for blockchain pruning")?
+                        .prune_blockchain(header.number, num_blocks_kept)
+                        .context("Pruning blockchain history")
```

Comment:
I'm wondering how this will behave while the sync thread is making progress and keeping a write SQLite transaction open.

Up to now, we had a _single_ writer to storage (the sync thread), all other tasksk did only reads. With this pruning task we now have _two_, so now it will be possible for `.transaction_with_behavior()` to just time out both in the sync thread and here. It seems like we're using the default 5s busy timeout, so _if_ the sync task does not complete the transaction within 5s (I think that's totally possible with slow storage and empty disk cache) we'll fail here. (Or vice versa: if we don't complete pruning in 5s the sync task can possibly time out.)

What is our strategy for handling that?